lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,671 @@
+// lib/Tranforms/IPO/StructFieldCacheAnalysis.cpp - Performs Cache-Aware
+// Structure Analysis-*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===------------------------------------------------------------------------===//
+//
+// This pass performs analysis on cache-aware structure field accesses based on
+// the following paper and reports recommendations on changes to make on the
+// source code to improve performance.
+//  [1] M. Hagog, C. Tice “Cache Aware Data Layout Reorganization Optimization
+//  in GCC”, Proceedings
+//      of the GCC Developers’ Summit,  Ottawa, 2005.
+//
+//===------------------------------------------------------------------------===//
+
+#include "StructFieldCacheAnalysisImpl.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Analysis/CFG.h"
+#include "llvm/Support/Format.h"
+
+#include <cmath>
+#include <queue>
+
+using namespace llvm;
+
+#define DEBUG_PRINT_COUNT(x) (format("%.2f", (x)))
+#define DEBUG_PRINT_DIST(x) (format("%.3f", (x)))
+
+static cl::opt<bool> PerformFRGOnly(
+    "struct-analysis-FRG-only", cl::init(false), cl::Hidden,
+    cl::desc("Stop the analysis after performing FRG generation"));
+
+static cl::opt<bool> PerformCPGCheck(
+    "struct-analysis-check-CPG", cl::init(false), cl::Hidden,
+    cl::desc("Perform CPG checking algorithm that takes a long time"));
+
+static cl::opt<bool> DisableIgnoreZeroCountNodes(
+    "struct-analysis-disable-ignore-zeros", cl::init(false), cl::Hidden,
+    cl::desc("Ignore zero-count nodes in FRG generation"));
+
+// Utility functions
```

Comment:
General comment about this file:  This file implements the really new/novel/critical parts of the analysis, so the algorithms used really need good comments.  Each function really needs a good comment explaining  what it's doing and why, and how it fits into the overall analysis.  This is really important, so that if someone new comes along in a year and wants to build on your code, they can figure out what's going on by reading the comments. (The *how* is obvious from the code, but the *why* is not).